### PR TITLE
fixing issue 738

### PR DESCRIPTION
**Pull request by yilongli**
fixed https://github.com/kframework/k/issues/738

### CODE REVIEW

**Comment by yilongli** [comment]
I could create a regression test to catch exactly this bug, but I am reluctant to do that because, in my opinion, we already have many small regression tests that are too weak to be useful. I believe the right way to avoid similar problems in the future is to have different checkers at the end of kompilation to validate certain properties of the kompiled definition. For example, in this case, we should have a checker saying that there should be no KLabels representing data structure constructors in the definition.

File: src/javasources/KTool/src/org/kframework/backend/java/rewritemachine/ExtendedSubstitution.java [hunk 1]
```diff
@@ -58,4 +58,9 @@ void setSubst(Map<Variable, Term> subst) {
     void addWriteCell(Cell<?> cell) {
         writeCells.add(cell);
     }
+
+    @Override
+    public String toString() {
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/rewritemachine/ExtendedSubstitution.java:63]: I don't understand how the changes to these three files relate to the change that fixes this issue.

Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/rewritemachine/ExtendedSubstitution.java:63]: having toString method makes debugging easy. I made the change during debugging.

**Comment by dwightguth** [comment]
Much of the problem with our small regression tests is that they are very slow to execute because they are essentially end-to-end tests. I suggest you make a JUnit unit test for CompileDataStructures which tests the basic functionality of the pass, including this use case. The last time I did something similar for one of the compiler passes, I found a number of latent bugs that had not even yet been detected with the code in question.

File: src/javasources/KTool/src/org/kframework/backend/java/rewritemachine/KAbstractRewriteMachine.java [hunk 2]
```diff
@@ -234,7 +234,15 @@ private Term getWriteCellRHS(String cellLabel) {
     }
 
     private Collection<Cell> getSubCellsByLabel(Cell<?> cell, String label) {
-        return ((CellCollection) cell.getContent()).cellMap().get(label);
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/rewritemachine/KAbstractRewriteMachine.java:237]: right okay fine so toString I kind of get. But what's this?

Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/rewritemachine/KAbstractRewriteMachine.java:237]: The bug is reported for Java backend without --pattern-matching option. I tried to test --pattern-matching as well and found a class cast exception.

**Comment by yilongli** [comment]
Jenkins: retest this please

**Comment by yilongli** [comment]
Writing a unit test to properly cover all use cases of `CompileDataStructures` is not easy and very time-consuming. Right now, I simply write a very specific unit test to catch the bug exposed in issue#738. I will probably come back some time to write unit tests for all important compilation passes (at the same time, develop some common infrastructure used frequently in building mock KIL objects).

**Comment by yilongli** [comment]
I think the time-out is due to the slight change in `CompileDataStructures` which results in a little more time spent in compilation. The time limit is already 40 mins. We should probably discuss how to reduce the build time effectively.

**Comment by yilongli** [comment]
@andreistefanescu please take a look at my changes

**Comment by andreistefanescu** [comment]
lgtm. I'm not sure what happens with the time. This change should not increase the compilation time by that much. Maybe one of the examples is not terminating now?

**Comment by yilongli** [comment]
Nope. The last example before time out is `bf` and it works on my laptop.

**Comment by dwightguth** [comment]
Jenkins: test this please

**Comment by dwightguth** [comment]
Next time this happens please ssh into the jenkins machine and run `top` to check the cpu usage. I am having trouble figuring out what triggers this, but sometimes processes get orphaned by jenkins and so the cpu gets hammered and builds time out. The fix is to just kill the processes. I will try to look into what is causing this today though to see if I can find a permanent solution.

**Comment by yilongli** [comment]
So after you kill the processes, the build time drops from 40mins to 24mins?

**Comment by dwightguth** [comment]
The reason it timed out was that it was trying to execute one process per thread when there were already processes from a previous build using up the cpus.

**Pull request merged**